### PR TITLE
Fix/oracle blocking

### PR DESCRIPTION
**Pull request by yan-soon**
- add getStateWithTimeout() so that oracle voting is not blocked during long software upgrades
- drop votes from oracle vote channel when size hits 1/2 of cap

### CODE REVIEW

File: oracle/reactor.go [hunk 1]
```diff
@@ -190,8 +190,17 @@ func (oracleR *Reactor) Receive(e p2p.Envelope) {
 		signatureWithoutPrefix, err := utils.GetSignatureWithoutPrefix(msg.Signature)
 		if err != nil {
 			logrus.Errorf("unable to get signature without prefix, invalid signature: %v", msg.Signature)
+			return
+		}
+
+		TIMEOUT := time.Second * 5
```
Comment (holyxiaoxin) [inline oracle/reactor.go]: Comment why 5seconds

File: oracle/service/runner/runner.go [hunk 2]
```diff
@@ -108,16 +115,22 @@ func PruneVoteBuffers(oracleInfo *types.OracleInfo, consensusState *cs.State) {
...
 		ticker := time.Tick(pruneInterval)
 		for range ticker {
-			lastBlockTime := consensusState.GetState().LastBlockTime.Unix()
-			currTimestampsLen := len(oracleInfo.BlockTimestamps)
-
-			if currTimestampsLen == 0 {
-				oracleInfo.BlockTimestamps = append(oracleInfo.BlockTimestamps, lastBlockTime)
-				continue
-			}
+			TIMEOUT := time.Second * 3
```
Comment (holyxiaoxin) [inline oracle/service/runner/runner.go]: Comment why 3s

File: oracle/reactor.go (hunk 1 above)
Comment (RavenXce) [inline oracle/reactor.go]: why is this caps

File: oracle/service/runner/runner.go (hunk 2 above)
Comment (RavenXce) [reply oracle/service/runner/runner.go]: 3s seems very low. why is the var caps?

File: oracle/service/runner/runner.go [hunk 3]
```diff
@@ -200,6 +213,11 @@ func Run(oracleInfo *types.OracleInfo, consensusState *cs.State) {
 			continue
 		}
 
+		// drop old votes when channel hits half cap
+		if len(oracleInfo.SignVotesChan) >= cap(oracleInfo.SignVotesChan)/2 {
```
Comment (RavenXce) [inline oracle/service/runner/runner.go:211]: why do we not log here?
why not just -1?

File: oracle/service/runner/runner.go [hunk 4]
```diff
@@ -108,16 +115,22 @@ func PruneVoteBuffers(oracleInfo *types.OracleInfo, consensusState *cs.State) {
...
-			}
+			TIMEOUT := time.Second * 3
+			chainState, err := consensusState.GetStateWithTimeout(TIMEOUT)
+			if err != nil {
+				log.Warnf("PruneVoteBuffers: timed out trying to get chain state after %v", TIMEOUT)
+			} else {
+				lastBlockTime := chainState.LastBlockTime.Unix()
+				currTimestampsLen := len(oracleInfo.BlockTimestamps)
+
+				if currTimestampsLen == 0 {
```
Comment (RavenXce) [inline oracle/service/runner/runner.go]: why is this not just `||` with L131

**Review by RavenXce — CHANGES_REQUESTED** [review]

File: oracle/service/runner/runner.go [hunk 5]
```diff
@@ -108,16 +108,18 @@ func PruneVoteBuffers(oracleInfo *types.OracleInfo, consensusState *cs.State) {
...
-
-			if currTimestampsLen == 0 {
-				oracleInfo.BlockTimestamps = append(oracleInfo.BlockTimestamps, lastBlockTime)
-				continue
-			}
-
-			if oracleInfo.BlockTimestamps[currTimestampsLen-1] != lastBlockTime {
-				oracleInfo.BlockTimestamps = append(oracleInfo.BlockTimestamps, lastBlockTime)
+			// keep this timeout close to lowest oracle resolution so that buffers do not build up even when chain is stale, but oracle service is still running
+			timeout := time.Second * 3
```
Comment (tehkohrz) [inline oracle/service/runner/runner.go]: change to prune interval

**Review by tehkohrz — CHANGES_REQUESTED** [review]

**Review by holyxiaoxin — APPROVED** [review]

**Pull request merged**